Source/SwiftLintFramework/Models/LinterCache.swift

```diff
@@ -110,31 +106,73 @@ public final class LinterCache {
         guard let url = location else {
             throw LinterCacheError.noLocation
         }
+        writeCacheLock.lock()
+        defer {
+            writeCacheLock.unlock()
+        }
         guard !writeCache.isEmpty else {
             return
         }
 
-        let cache = mergeCaches()
-        let json = try cache.toJSON()
-        try json.write(to: url, atomically: true, encoding: .utf8)
+        readCacheLock.lock()
+        let readCache = lazyReadCache
+        readCacheLock.unlock()
+
+        for (description, writeFileCache) in writeCache {
+            guard !writeFileCache.isEmpty else {
+                continue
+            }
+
+            let fileCache = readCache[description]?.merging(writeFileCache) { _, write in write } ?? writeFileCache
+            let json = try fileCache.toJSON()
+            let file = url.appendingPathComponent(description).appendingPathExtension("json")
+            try json.write(to: file, options: .atomic)
+        }
     }
 
     internal func flushed() -> LinterCache {
         return LinterCache(cache: mergeCaches(), location: location,
                            fileManager: fileManager, swiftVersion: swiftVersion)
     }
 
+    private func fileCache(cacheDescription: String) -> FileCache {
+        readCacheLock.lock()
+        defer {
+            readCacheLock.unlock()
+        }
+        if let fileCache = lazyReadCache[cacheDescription] {
+            return fileCache
+        }
+
+        guard let location = location else {
+            return [:]
+        }
+
+        let file = location.appendingPathComponent(cacheDescription).appendingPathExtension("json")
```

Comment:
Will `appendingPathComponent(cacheDescription)` always produce a safe path for writing? I don't remember how we compute it these days (we've changed it over time) so I'll need to trace through the code to confirm.

Reply:
It's an `MD5`, that should be safe.